### PR TITLE
[SPARK-48545][SQL] Create to_avro and from_avro SQL functions to match DataFrame equivalents

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR creates two new SQL functions "to_avro" and "from_avro" to match existing DataFrame equivalents.

For example:

```
sql(
  """
    |create table t as
    |  select named_struct('u', named_struct('member0', member0, 'member1', member1)) as s
    |  from values (1, null), (null,  'a') tab(member0, member1)
    |""".stripMargin)

val jsonFormatSchema =
  """
    |{
    |  "type": "record",
    |  "name": "struct",
    |  "fields": [{
    |    "name": "u",
    |    "type": ["int","string"]
    |  }]
    |}
    |""".stripMargin

spark.sql(
  s"""
    |select from_avro(result, '$jsonFormatSchema', map()).u from (
    |  select to_avro(s, '$jsonFormatSchema') as result from t
    |)")
  .collect()

> {1, NULL}
  {NULL, "a"}
```

### Why are the changes needed?

This brings parity between SQL and DataFrame APIs in Apache Spark.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds extra unit tests, and I also checked that the functions work with `spark-shell`.

### Was this patch authored or co-authored using generative AI tooling?

No GitHub copilot usage this time

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    > SELECT IS_NULL(_FUNC_(result, '{"type": "record", "name": "struct", "fields": [{ "name": "u", "type": ["int","string"] }]}')) AS result FROM (SELECT NAMED_STRUCT('u', NAMED_STRUCT('member0', member0, 'member1', member1)) AS s FROM VALUES (1, NULL), (NULL,  'a') tab(member0, member1));
+     [false]
+  """,
+  note = """
+    The specified schema must match actual schema of the read data, otherwise the behavior
+    is undefined: it may fail or return arbitrary result.
+    To deserialize the data with a compatible and evolved schema, the expected Avro schema can be
+    set via the corresponding option.
+  """,
+  group = "misc_funcs",
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:50]: Do we want to have a new group for avro functions? Like we have `csv_funcs` and `json_funcs`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  since = "4.0.0"
+)
+case class FromAvroSqlFunction(child: Expression, jsonFormatSchema: Expression, options: Expression)
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:53]: How about simply calling it `FromAvro` (and `ToAvro`)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala]: Sounds good, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala]: Sounds good, done.

**Comment by dtenedor** [comment]
Thanks @allisonwang-db for your review, followed through on your comments.

**Review by allisonwang-db — APPROVED** [review]
Looks good! cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]
LGTM if CI passes

**Comment by dtenedor** [comment]
cc @cloud-fan the CI is passing now :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:1]: to => To ?

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:1]: good Q; I copied the convention of existing files with multiple expressions in them with a leading lowercase character such as [1]

[1] sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    val optionsValue: Map[String, String] = options.eval() match {
+      case a: ArrayBasedMapData if a.keyArray.array.nonEmpty =>
+        val keys: Array[String] = a.keyArray.array.map(_.toString)
+        val values: Array[String] = a.valueArray.array.map(_.toString)
+        keys.zip(values).toMap
+      case _ =>
+        Map.empty
+    }
+    val constructor =
+      Utils.classForName("org.apache.spark.sql.avro.AvroDataToCatalyst").getConstructors().head
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:111]: Nit: Avro data source is not loaded by default. If the new function is called and avro package is not loaded, we should show user-friendly error messages

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:1]: I see. Let's keep this unchanged then.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala (hunk 3 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/toFromAvroSqlFunctions.scala:111]: Sounds good, can help with this in a follow=up :)

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**